### PR TITLE
Generic algorithm to create a shallow, memberwise clone of a node.

### PR DESCRIPTION
**Pull request by rbuckton**
Eventually will be needed for transformations.

### CODE REVIEW

File: src/compiler/utilities.ts [hunk 1]
```diff
@@ -1580,20 +1580,66 @@ namespace ts {
...
+     * @param location An optional TextRange to use to supply the new position.
+     * @param flags The NodeFlags to use for the cloned node.
+     * @param parent The parent for the new node.
+     */
+    export function cloneNode<T extends Node>(node: T, location?: TextRange, flags?: NodeFlags, parent?: Node): T {
+        // We don't use "clone" from core.ts here, as we need to preserve the prototype chain of
+        // the original node. We also need to exclude specific properties and only include own-
+        // properties (to skip members already defined on the shared prototype).
+        const clone = location !== undefined
+            ? <T>createNode(node.kind, location.pos, location.end)
```
Comment (sandersn) [inline src/compiler/utilities.ts]: kind is copied again in the for loop, right? I just want to make sure I'm reading the code correctly.

**Comment by sandersn** [comment]
Looks good to me.

File: src/compiler/utilities.ts (hunk 1 above)
Comment (rbuckton) [inline src/compiler/utilities.ts]: Good catch. I originally wrote the `cloneNode` function before @vladima modified the Node constructor, when `kind` was set on the prototype. I can probably drop `isExcludedPropertyForClone` in favor of `clone.hasOwnProperty(key)`.

Comment (sandersn) [inline src/compiler/utilities.ts]: This will copy over `flags` and `parent` now, won't it? Which is bad unless they are provided by the caller.

Comment (rbuckton) [inline src/compiler/utilities.ts]: No, flags and parent are set in the Node constructor.  See line 808 of core.ts:

```ts
    function Node(kind: SyntaxKind, pos: number, end: number) {
        this.kind = kind;
        this.pos = pos;
        this.end = end;
        this.flags = NodeFlags.None;
        this.parent = undefined;
    }
```

Comment (sandersn) [inline src/compiler/utilities.ts]: All right, sounds good.

**Pull request merged**